gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java

```diff
@@ -278,18 +281,34 @@ protected String getTimestampDisplay(long timestamp) {
    * Method that deletes expired tokens based on the token timestamp.
    */
   protected void evictExpiredTokens() {
-    try {
-      for (final String token : getTokens()) {
-        if (isExpired(token)) {
+    for (final String token : getTokens()) {
+      try {
+        if (needsEviction(token)) {
           log.evictToken(getTokenDisplayText(token));
           removeToken(token);
         }
+      } catch (final Exception e) {
+        log.failedExpiredTokenEviction(getTokenDisplayText(token), e);
       }
-    } catch(final Exception e) {
-      log.failedExpiredTokenEviction(e);
     }
   }
 
+  /**
+   * Method that checks if an expired token is ready to be evicted
+   * by adding configured grace period to the expiry time.
+   * @param token
+   * @return
+   */
+  protected boolean needsEviction(final String token) {
+    boolean needsEviction;
+    needsEviction = isUnknown(token); // Check if the token exist
+    if (!needsEviction) {
+      /* If it not unknown, check if it is expired and within grace period before evicting */
+      needsEviction = (getTokenExpiration(token) + TimeUnit.SECONDS.toMillis(tokenEvictionGracePeriod)) <= System.currentTimeMillis();
```

Comment:
I think this whole method could be simply this single statement:
return ((getTokenExpiration(token) + TimeUnit.SECONDS.toMillis(tokenEvictionGracePeriod)) <= System.currentTimeMillis());